src/modules/languages/rust.nix

```diff
@@ -13,67 +10,91 @@ let
+        List of [Rustup components](https://rust-lang.github.io/rustup/concepts/components.html)
+        to install. Defaults to those available in `nixpkgs`.
+      '';
+    };
+
+    channel = lib.mkOption {
+      type = lib.types.enum [ "nixpkgs" "stable" "beta" "nightly" ];
+      default = "nixpkgs";
+      defaultText = lib.literalExpression ''"nixpkgs"'';
+      description = "The rustup toolchain to install.";
+    };
+
+    toolchain = lib.mkOption {
       type = lib.types.submodule ({
-        options = {
-          rust-src = lib.mkOption {
-            type = lib.types.either lib.types.package lib.types.str;
-            default = pkgs.rustPlatform.rustLibSrc;
-            defaultText = lib.literalExpression "pkgs.rustPlatform.rustLibSrc";
-            description = "rust-src package";
-          };
-        }
-        // genAttrs tools (name: lib.mkOption {
-          type = lib.types.package;
-          default = pkgs.${name};
-          defaultText = lib.literalExpression "pkgs.${name}";
-          description = "${name} package";
-        });
+        freeformType = lib.types.attrsOf lib.types.package;
+
+        options =
+          let
+            documented-components = [ "rustc" "cargo" "clippy" "rustfmt" "rust-analyzer" ];
+            mkComponentOption = component: lib.mkOption {
+              type = lib.types.nullOr lib.types.package;
+              default = pkgs.${component};
+              defaultText = lib.literalExpression "pkgs.${component}";
+              description = "${component} package";
+            };
+          in
+          lib.genAttrs documented-components mkComponentOption;
       });
-      defaultText = lib.literalExpression "pkgs";
       default = { };
-      description = "Attribute set of packages including rustc and Cargo.";
-    };
-
-    version = lib.mkOption {
-      type = lib.types.nullOr lib.types.str;
-      default = null;
-      description = "Set to stable, beta, or latest.";
+      defaultText = lib.literalExpression "nixpkgs";
+      description = "Rust component packages. May optionally define additional components, for example `miri`.";
     };
   };
 
   config = lib.mkMerge [
     (lib.mkIf cfg.enable {
-      packages = attrValues (getAttrs tools cfg.packages) ++ lib.optional pkgs.stdenv.isDarwin pkgs.libiconv;
+      packages = (builtins.map (c: cfg.toolchain.${c} or (throw (error "toolchain.${c}"))) cfg.components)
+        ++ lib.optional pkgs.stdenv.isDarwin pkgs.libiconv;
 
       # enable compiler tooling by default to expose things like cc
       languages.c.enable = lib.mkDefault true;
 
-      env.RUST_SRC_PATH = cfg.packages.rust-src;
+      # RUST_SRC_PATH is necessary when rust-src is not at the same location as
+      # as rustc. This is the case with the rust toolchain from nixpkgs.
+      env.RUST_SRC_PATH =
+        if cfg.toolchain ? rust-src
+        then "${cfg.toolchain.rust-src}/lib/rustlib/src/rust/library"
+        else pkgs.rustPlatform.rustLibSrc;
```

Comment:
If it makes things easier, we only really need this for "nixpkgs".

Reply:
> we only really need this for "nixpkgs"

i don't think that's true:
```nix
enable = true;
toolchain.rust-std = my-fenix-toolchain.rust-std;
```

Reply:
Actually I just checked the docs so I may be wrong on this. Mixed it up with rust-overlay. I would leave it as is.

Reply:
for clarity, when you say "leave it as is" you mean my current if-expression is fine?

Reply:
It's `rust-src` in this instance. But yeah, I may be wrong here, sorry.